dronegen/tag.go

```diff
@@ -0,0 +1,425 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	// rpmPackage is the RPM package type
+	rpmPackage = "rpm"
+	// debPackage is the DEB package type
+	debPackage = "deb"
+)
+
+type tagBuildType struct {
+	os      string
+	arch    string
+	fips    bool
+	centos6 bool
+}
+
+// tagMakefileTarget gets the correct tag pipeline Makefile target for a given arch/fips combo
+func tagMakefileTarget(params tagBuildType) string {
+	makefileTarget := fmt.Sprintf("release-%s", params.arch)
+	if params.centos6 {
+		makefileTarget += "-centos6"
+	}
+	if params.fips {
+		makefileTarget += "-fips"
+	}
+	return makefileTarget
+}
+
+// tagCheckoutCommands builds a list of commands for Drone to check out a git commit on a tag build
+func tagCheckoutCommands(fips bool) []string {
+	commands := []string{
+		`mkdir -p /go/src/github.com/gravitational/teleport`,
+		`cd /go/src/github.com/gravitational/teleport`,
+		`git clone https://github.com/gravitational/${DRONE_REPO_NAME}.git .`,
+		`git checkout ${DRONE_TAG:-$DRONE_COMMIT}`,
+		// fetch enterprise submodules
+		`mkdir -m 0700 /root/.ssh && echo -n "$GITHUB_PRIVATE_KEY" > /root/.ssh/id_rsa && chmod 600 /root/.ssh/id_rsa`,
+		`ssh-keyscan -H github.com > /root/.ssh/known_hosts 2>/dev/null && chmod 600 /root/.ssh/known_hosts`,
+		`git submodule update --init e`,
+		// this is allowed to fail because pre-4.3 Teleport versions don't use the webassets submodule
+		`git submodule update --init --recursive webassets || true`,
+		`rm -f /root/.ssh/id_rsa`,
+		// create necessary directories
+		`mkdir -p /go/cache /go/artifacts`,
+		// set version
+		`if [[ "${DRONE_TAG}" != "" ]]; then echo "${DRONE_TAG##v}" > /go/.version.txt; else egrep ^VERSION Makefile | cut -d= -f2 > /go/.version.txt; fi; cat /go/.version.txt`,
+	}
+	return commands
+}
+
+// tagBuildCommands generates a list of commands for Drone to build an artifact as part of a tag build
+func tagBuildCommands(params tagBuildType) []string {
+	commands := []string{
+		`apk add --no-cache make`,
+		`chown -R $UID:$GID /go`,
+		`cd /go/src/github.com/gravitational/teleport`,
+	}
+
+	if params.fips {
+		commands = append(commands,
+			"export VERSION=$(cat /go/.version.txt)",
+		)
+	}
```

Comment:
curious - why is this only for FIPS builds?

Reply:
It's a byproduct of the fact that `e/Makefile` doesn't set `VERSION` itself; it relies on it being inherited from the OSS `Makefile` in the root directory because `e` is intended to be a submodule and doesn't have a way to set `VERSION` as a standalone checkout. As FIPS builds are Enterprise-only the variable doesn't get inherited.

We could actually set `VERSION` using this method for every single build with no negative consequences, but I'm trying to avoid introducing additional changes as part of the `dronegen` migration. We can look to make improvements later.